libs/platform/ebpf_epoch.c

```diff
@@ -10,93 +10,132 @@
-// If the stale flag is already set, then the per-CPU stale_worker is scheduled.
-//
-// Thread entry table:
-// The thread entry table is a fixed size per-CPU hash table for tracking per-thread ebpf_epoch_state_t. The hash is
-// based on the thread ID. The thread entry table is protected by a semaphore that limits the number of threads
-// that can be active in the epoch at any one time. If a thread attempts to enter the epoch and the semaphore
-// count is 0, then the thread blocks until a thread exits the epoch, which ensures that the thread entry table
-// has an available entry for the thread to use.
+// 4. Set the _ebpf_epoch_cpu_entry_t.rundown_in_progress flag to true and send the message to the next CPU.
+// If this is the last CPU, then signal the caller that rundown is complete.
 
 // Delay after the _ebpf_flush_timer is set before it runs.
-#define EBPF_EPOCH_FLUSH_DELAY_IN_MICROSECONDS 1000
-
-// Time before logging that a thread entry is stale
-#define EBPF_EPOCH_STALE_THREAD_TIME_IN_NANO_SECONDS 10000000000 // 10 seconds
-
-// The maximum time a thread remains in the thread table while inactive.
-#define EBPF_EPOCH_THREAD_TABLE_TIMEOUT_IN_NANO_SECONDS 1000000000 // 1 second
+#define EBPF_EPOCH_FLUSH_DELAY_IN_NANOSECONDS 1000000
 
 #define EBPF_NANO_SECONDS_PER_FILETIME_TICK 100
 
-typedef struct _ebpf_epoch_state
-{
-    int64_t is_preemptible : 1; // Is the thread preemptible?
-    int64_t epoch : 63;         // The highest epoch seen by this epoch state.
-} ebpf_epoch_state_t;
-
-#define EBPF_EPOCH_THREAD_TABLE_SIZE \
-    (EBPF_CACHE_LINE_SIZE / sizeof(ebpf_epoch_state_t)) // Number of entries in the thread table.
-
-// Number of reserved entries in the thread table.
-// One for the current thread running at DISPATCH_LEVEL.
-#define EBPF_EPOCH_THREAD_TABLE_RESERVED_COUNT 1
-
 // Table to track per CPU state.
-// This table must fit into a multiple of EBPF_CACHE_LINE_SIZE.
+// Each entry is only accessed by the CPU that owns it and only at IRQL >= DISPATCH_LEVEL.
+// This ensures that no locks are required to access the per CPU state.
 #pragma warning(disable : 4324) // Structure was padded due to alignment specifier.
 typedef __declspec(align(EBPF_CACHE_LINE_SIZE)) struct _ebpf_epoch_cpu_entry
 {
-    _Guarded_by_(lock) ebpf_epoch_state_t
-        epoch_table[EBPF_EPOCH_THREAD_TABLE_SIZE]; // Epochs on this CPU. If zero, then it is not active.
-    ebpf_lock_t lock;
-    KSEMAPHORE* epoch_table_semaphore;              // Semaphore to number of threads active in the epoch per CPU.
-    _Guarded_by_(lock) KDPC* stale_worker;          // Per-CPU stale worker DPC.
-    _Guarded_by_(lock) ebpf_list_entry_t free_list; // Per-CPU free list.
-    _Guarded_by_(lock) int timer_armed : 1;
-    _Guarded_by_(lock) int stale : 1;
-    _Guarded_by_(lock) int rundown_in_progress : 1;
+    LIST_ENTRY thread_list;                ///< Per-CPU list of thread entries.
```

Comment:
its a list of epoch_states; may be rename it as such?

Reply:
Done.